lib/Optimizer/Lower.cpp

```diff
@@ -647,6 +647,93 @@ static void lowerGroupConvolutionNode(Function *F, CompilationContext &cctx,
 }
 
+static void lowerChannelwiseQuantizedConvolutionNode(
+    Function *F, CompilationContext &cctx,
+    const ChannelwiseQuantizedConvolutionNode &CQC) {
+  // ChannelwiseQuantizedConvolutionNode can be represented as a Concatenation
+  // of smaller dimension quantized Convolutions each with their own qparams.
+  // Input channels will be divided into equal groups of consecutive channels.
+  // These will be separately convolved each with its own filter and bias.
+  // This will result in 4 * Group + 1 nodes.
+
+  // Only lowering of groupwise, not channelwise is supported so far.
+  if (!CQC.getGroupwise()) {
+    return;
+  }
+
+  llvm::ArrayRef<unsigned_t> kernels = CQC.getKernels();
+  llvm::ArrayRef<unsigned_t> pads = CQC.getPads();
+  llvm::ArrayRef<unsigned_t> strides = CQC.getStrides();
+  unsigned_t group = CQC.getGroup();
+  auto in = CQC.getInput();
+
+  Constant *filter = llvm::cast<Constant>(CQC.getFilter());
+  Constant *bias = llvm::cast<Constant>(CQC.getBias());
+  Constant *scales = llvm::cast<Constant>(CQC.getScales());
+  Constant *offsets = llvm::cast<Constant>(CQC.getOffsets());
+
+  ShapeNHWC idim = ShapeNHWC(in.dims());
+  ShapeHW kdim(kernels);
+  unsigned inCperG = idim.c / group;
+  unsigned outCperG = filter->dims()[0] / group;
+
+  auto convOutDims = CQC.getResult().dims().vec();
+  convOutDims[3] = outCperG;
+
+  auto filterDims = filter->dims().vec();
+  filterDims[0] = outCperG;
+  filterDims[3] = inCperG;
+
+  // Final output type of each convolution after rescaling
+  auto finalConvOutTy = F->getParent()->uniqueTypeWithNewShape(
+      CQC.getResult().getType(), convOutDims);
+
+  auto scalesHandle = scales->getHandle<float>();
+  auto offsetsHandle = offsets->getHandle<int32_t>();
+
+  Module *M = F->getParent();
+
+  std::vector<NodeValue> branches;
+  for (unsigned_t groupId = 0; groupId < group; groupId++) {
+    float filterScale = scalesHandle.raw(groupId);
+    int32_t filterOffset = offsetsHandle.raw(groupId);
+
+    SliceNode *inSlice =
+        F->createSlice(CQC.getName(), in, {0, 0, 0, groupId * inCperG},
+                       {idim.n, idim.h, idim.w, (groupId + 1) * inCperG});
+
+    // Create quantized filter sliced Constant.
+    Tensor slicedFilterTensor = filter->getPayload().getOwnedSlice(
+        filterDims, {outCperG * groupId, 0, 0, 0});
+    auto quantizedFilterType = slicedFilterTensor.getType();
+    quantizedFilterType.scale_ = filterScale;
+    quantizedFilterType.offset_ = filterOffset;
+    slicedFilterTensor.setType(&quantizedFilterType);
+    Constant *slicedFilterConst = M->createConstant(
+        strFormat("%s_filter", CQC.getName().data()), slicedFilterTensor);
+
+    // Create bias sliced Constant. Bias's scale should always be inputScale *
+    // filterScale.
+    TensorQuantizationParams tqp;
+    tqp.offset = 0;
+    tqp.scale = inSlice->getInput().getType()->getScale() * filterScale;
+    Tensor slicedBiasTensor =
```

Comment:
This feels like something we should add as a graph optimization (i.e. `Slice(Constant) -> Constant`). Then you could just add Slice/Quantize nodes here.

Reply:
I don't think we have it yet -- I think we have it for Splat but not Constant. Don't need it for this PR necessarily, but perhaps file an issue for a new opt and mention there that we could then remove the slicing here in the lowering pass so we don't forget.

Reply:
Looks like it already exists https://github.com/pytorch/glow/issues/2168